homeassistant/loader.py

```diff
@@ -506,31 +506,35 @@ async def async_get_integration(hass: HomeAssistant, domain: str) -> Integration
 
     event = cache[domain] = asyncio.Event()
 
+    try:
+        integration = await _async_get_integration(hass, domain)
+    except Exception:  # pylint: disable=broad-except
+        # Remove event from cache.
+        cache.pop(domain)
+        event.set()
+        raise
+
+    cache[domain] = integration
+    event.set()
+    return integration
+
+
+async def _async_get_integration(hass: HomeAssistant, domain: str) -> Integration:
     # Instead of using resolve_from_root we use the cache of custom
     # components to find the integration.
-    integration = (await async_get_custom_components(hass)).get(domain)
-    if integration is not None:
-        validate_custom_integration_version(integration)
+    if integration := (await async_get_custom_components(hass)).get(domain):
         _LOGGER.warning(CUSTOM_WARNING, integration.domain)
-        cache[domain] = integration
-        event.set()
+        validate_custom_integration_version(integration)
```

Comment:
Shouldn't this be before the warning?
If it is not loaded there is no need to print this warning?

Reply:
It seemed relevant either way 🤷‍♂️ but I see your point.

adjusted in 28e40e7918